Move ferc eia manual mapping notebook

aesharpe opened the pull request:
<!--

Making a PUDL Pull Request

Before making a PR you may want to check out our:

* contributing guidelines: https://catalystcoop-pudl.readthedocs.io/en/latest/CONTRIBUTING.html
* code of conduct: https://catalystcoop-pudl.readthedocs.io/en/latest/code_of_conduct.html
* development process: https://catalystcoop-pudl.readthedocs.io/en/latest/dev/index.html

## PR Process Overview

* PRs have to get an approving review before merging into their development branch.
* Most PRs should be made against the `dev` branch, unless they are part of some larger ongoing refactoring, in which case there will be a persistent development branch for that work.
* It is much easier to do timely code reviews on smaller chunks of code. We try to keep PRs under 500 lines of code.
* Draft PRs are a good way to get early feedback on designs or several incremental commits that will add up to larger changes. If you want a review of a draft PR, make sure you contact the reviewer directly or mention their username in the PR comment, so they get a notification.
* How quickly we can review a PR will depend on how large and complex it is, and how busy we are, but ideally we strive to get an initial review done within a week. If there are going to be delays, we should at least comment on the PR to let you know the situation.
* If you believe you've addressed a reviewer's comments, respond with a brief note and mark the comment resolved. If further discussion is requried respond and do not resolve the comment.
* Before a PR is merged all reviewer comments should be resolved. If a reviewer doesn't feel that their comment has been sufficiently addressed, they may unresolve a comment.
* Be careful not to accidentally "start a review" when responding to comments! If this does happen, don't forget to submit the review you've started so the other PR participatns can see your comments (they are invisible to others if marked "Pending").
* In the period after an initial review when there is significant back-and-forth with the reviewer deciding what changes should actually be made, there should probably be daily interaction. If significant changes are required, it's usually best to request another review after those changes have been made.

Feel free to delete the commented-out parts of the template before submitting the PR.

-->

# PR Overview

This PR creates a new directory in `devtools/ferc1-eia-glue` called `training_data` where the notebooks for running the manual mapping spreadsheet creation and validation and their outputs are housed. 

Previously in the `rmi-ferc1-eia` repo, the `create_override_spreadsheets.py` module now lives in **`pudl/analysis/ferc1_eia_train.py`**. I tweaked some of the functions so they have updated paths. I also fixed the `allow_override_overrides` parameter so that if true it overrides duplicate training data with the newer version you submit (the older version would only work if there were `record_id_ferc1` and `record_id_eia` duplicates, but there are many different versions of a `record_id_eia` thanks to the plant parts table. Now the param only needs `record_id_ferc1` to be duplicated in order to override. This prevents duplicate training data which causes the `ferc1_eia` analysis that relies on the training data to fail.

This PR also takes the `make_overrides.ipynb` notebook from the rmi repo and moves it into the `devtools/ferc1-eia-glue/training_data` directory. It also splits the contents of the notebook into two: 
- **`create_FERC1-EIA_manual_mapping_spreadsheets.ipynb`** for making the override sheets. Outputs from this process (the fresh override spreadsheets) live in the new `devtools/ferc1-eia-glue/training_data/manual_mapping_spreadsheets` directory.
- **`validate_and_integrate_FERC1-EIA_manually_mapped_records.ipynb`** for validating and integrating the manually mapped FERC-EIA records. The inputs for this process (mapped data you want to test/integrate) must be put in the new `devtools/ferc1-eia-glue/training_data/add_to_training` directory.

Both of these new directories `add_to_training` and `manual_mapping_spreadsheets` are empty unless you either put things in there in the case of `add_to_training` or run the `create_FERC1-EIA_manual_mapping_spreadsheets.ipynb` in the case of `manual_mapping_spreadsheets`.
 
These functions aren't the most well designed, and they aren't really permanent, so I didn't bother cleaning them up toooo much. 

I also added the training data in for Southern, IDACORP, and Evergy

<!--

Include a short narrative summary of what's going on in the PR. This can be a bulleted list. You might want to include:

* What are you changing and why?
* Are there any known unsolved problems remaining in the PR?
* Is there anything that you want a reivewer to pay particular attention to?
* What kind of feedback are you looking for on the PR?

-->

# PR Checklist

- [ ] Merge the most recent version of the branch you are merging into (probably `dev`).
- [ ] All CI checks are passing. [Run tests locally to debug failures](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)
- [ ] Make sure you've included good docstrings.
- [ ] For major data coverage & analysis changes, [run data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation)
- [ ] Include unit tests for new functions and classes.
- [ ] Defensive data quality/sanity checks in analyses & data processing functions.
- [ ] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/latest/release_notes.html) and reference reference the PR and related issues.
- [ ] Do your own explanatory review of the PR to help the reviewer understand what's going on and identify issues preemptively.

zaneselvans commented:
Should this maybe go under `devtools`? Is it the kind of thing that we're going to need to pull out periodically and update like we do with the FERC-EIA plant IDs or the other ETL debugging/development notebooks?  The stuff under `notebooks` is a bit of a mess.

cmgosnell reviewed (changes requested):
I added some suggestions for where some of these files/code could live in the comments. I added some [notebook comments as well](https://app.reviewnb.com/catalyst-cooperative/pudl/blob/move-ferc-eia-manual-mapping-notebook/notebooks%2Fwork-in-progress%2Fmake_ferc_eia_training_data.ipynb%2F/discussion) (idk if they will show up over here so i wanted to point them out).

While i'm trying to integrate the duke training data, I'm getting a new error when trying to run `ferc1_eia_df = pudl_out.ferc1_eia()` on all of the years about duplicate records.

```
File ~/code/pudl/src/pudl/analysis/ferc1_eia.py:501, in run_model(features_train, features_all, train_df)
    499 lrc = LogisticRegression()
    500 clf = GridSearchCV(estimator=lrc, param_grid=param_grid, verbose=True, n_jobs=-1)
--> 501 clf.fit(X=x_train, y=y_train)
...
ValueError: Found input variables with inconsistent numbers of samples: [6903, 6942]
```

cmgosnell commented on src/pudl/analysis/ferc1_eia.py:
```diff
@@ -1012,3 +1013,691 @@ def add_null_overrides(connects_ferc1_eia):
     return connects_ferc1_eia
+
+
+#######################################################################################
+# Create Training Data / Override Tools
+#######################################################################################
+# """
+# Create output spreadsheets used for overriding and checking FERC-EIA record matches.
+
+# The connect_ferc1_to_eia.py module uses machine learning to link records from FERC
+# with records from EIA. While this process is waaay better more efficient and logical
+# than a human, it requires a set of hand-compiled training data in order to do it's job.
+# On top of that, it's not always capable of mapping some of the trickier records that
+# humans might be able to make sense of based on past or surrounding records.
+
+# This module creates an output spreadsheet, based on a certain utility, that makes the
+# matching and machine-matched human validation process much easier. It also contains
+# functions that will read those new/updated/validated matches from the spreadsheet,
+# validate them, and incorporate them into the existing training data.
+
+# """
```
I honestly think the two steps (being 1. make the spreadsheets for manual mapping & 2. integrate those spreadsheets into the training data) could be a little more separated. It may be easier to have two modules and/or two notebooks. maybe even just the two notebooks seem better.

aesharpe replied:
Ugh, I agree, I just wasn't sure how much time it was worth spending on this. I think we should have a convo about what this process will look like moving forward before refactoring.

aesharpe replied:
I could also make step 1 into a script?

aesharpe replied:
I could also make the validation step into a script...

aesharpe replied:
Verdict: not making them into scripts for time. Just made two separate notebooks

aesharpe commented:
> Should this maybe go under `devtools`? Is it the kind of thing that we're going to need to pull out periodically and update like we do with the FERC-EIA plant IDs or the other ETL debugging/development notebooks? The stuff under `notebooks` is a bit of a mess.

That's a good point.

aesharpe commented:
Question for @cmgosnell: If you checkout this branch, do you see the directories `devtools/ferc1-eia-glue/training_data/add_to_training` and `devtools/ferc1-eia-glue/training_data/manual_mapping_spreadsheets`?? I feel like they're not included here. I'm not sure how to add empty directories to a PR.

zaneselvans commented:
Git won't track empty directories, only files. There needs to be a file of some kind in the directory if you want the directory to show up. If there are frequently going to be files in these directories that should **not** be tracked by git, then a good option might be to add a `.gitignore` to each of them that specifies what types of files should be ignored (e.g. `*.csv`).

cmgosnell commented on src/pudl/analysis/ferc1_eia_train.py:
```diff
@@ -0,0 +1,712 @@
+    ), f"{id_col} {error_message}: {bad_ids}"
+
+
+def already_in_training(training_data, validated_connections):
+    """Blah."""
+    logger.info("checking if already in validation or not")
+    training_vals = training_data.record_id_eia.dropna().unique().tolist()
+    val_vals = validated_connections.record_id_eia_override_1.dropna().unique().tolist()
+
+    not_in_training = [x for x in val_vals if x not in training_vals]
+    print(f"Total records: {len(val_vals)}")
+    print(f"Records not in training data: {len(not_in_training)}")
+    print(not_in_training)
+    print("")
```
what does this actually produce? in the notebook it just spits out `not_in_training` without a real indication of what those records are or what I'm supposed to do with them

aesharpe replied:
I'll clarify the logger -- it was a last minute thing I added so that I could tell if our manual mapping spreadsheets had already been integrated into the training data or not. Mostly an artifact of this process taking months / years to complete and old mapping files going stale.

aesharpe replied:
Actually, I clarified the doc string and removed it from the notebook, because it's only really necessary on specific occasions (when you forget what you've done...)

aesharpe commented:
> there shouldn't be any matches between FERC and EIA
that have EIA matches but aren't in the FERC plant table, but we
got 186.

What exactly does this mean? Like bad ferc1 ids?

zaneselvans commented:
@aesharpe If a directory contains no tracked files, the directory is not tracked.  But if you put a `.gitignore` in the directory, the `.gitignore` file will be tracked, and so the directory will be tracked.

aesharpe commented:
> in the validate and integrate notebook, I like that you added space for someone to add their own record id, but this means you can't restart and run all cells to check if its actually working. my suggestion is to just add a random id and keep the note in there.

Fixed!

> in the validate and integrate notebook you renamed training_df to current_training_df but didn't update it in future cells.

Fixed!

aesharpe commented:
I'm testing some tests that I made and running into this error that comes from the `ferc1_eia.py` module: 

```
        not_in_ppe = train_df[train_df._merge == "left_only"]
        if not not_in_ppe.empty:
>           raise AssertionError(
                "Not all training data is associated with EIA records.\n"
                "record_id_ferc1's of bad training data records are: "
                f"{list(not_in_ppe.reset_index().record_id_ferc1)}"
            )
E           AssertionError: Not all training data is associated with EIA records.
E           record_id_ferc1's of bad training data records are: ['f1_steam_2012_12_57_5_1', 'f1_steam_2013_12_57_5_1', 'f1_steam_2005_12_186_3_5', 'f1_steam_2006_12_186_3_5', 'f1_steam_2007_12_186_3_5', 'f1_steam_2008_12_186_3_5', 'f1_steam_2009_12_186_3_5', 'f1_steam_2010_12_186_3_5', 'f1_steam_2011_12_186_3_5', 'f1_steam_2012_12_186_3_5', 'f1_steam_2013_12_186_3_5', 'f1_steam_2005_12_159_4_1', 'f1_steam_2006_12_159_4_1', 'f1_steam_2007_12_159_3_5', 'f1_steam_2008_12_159_3_5', 'f1_steam_2009_12_159_3_5', 'f1_steam_2010_12_159_3_5', 'f1_steam_2011_12_159_3_5', 'f1_steam_2012_12_159_3_5', 'f1_steam_2013_12_159_3_5']
```

I checked these records in the training data and their corresponding EIA records are in the PPL. @cmgosnell do you know what might be going on?

cmgosnell commented:
I have a suspicion! Are they plant_unit's? and are they failing on the integration test with only the two years of data?

If so: this the `unit_id_pudl` is not stable depending on the years of data being loaded. This stinks and is a bug and means we need to be a little more careful about record_id_eia's for units.
for now:
* for any units you can go in and find another ppe that is probably labeled as `true_gran == False` but has the same `appro_record_id_eia == {whatever your unit record id is}` and use that one.
for medium:
* we could do this automatically to all of the training data. i.e. convert all `plant_part == "plant_unit"` to another record that is the same granularity
for forever:
* we should convert the record_id for the units in the ppe. I've been reluctant to do this because i've known we have all of these manual mapping sheets lying around & we'd either have to change all of the IDs or have the units be generated in a different way.

aesharpe commented:
> Are they plant_unit's?

Most of them appear to be like this: `3295_natural_gas_fired_combined_cycle_2005_plant_technology_total_17539`

and have the natural gas fired combined cycle tag

cmgosnell commented:
@aesharpe are these the failures on the CI? if your trying to validation *all* of the training data, there will be old years that don't show up in the `ppe` because it doesn't have all of the years. You can use `pudl.analysis.ferc1_eia.restrict_train_connections_on_date_range`

aesharpe commented on test/integration/ferc1_eia_train_test.py:
```diff
@@ -0,0 +1,116 @@
+            ["4270_2020_plant_total_11241"],
+            ["f1_steam_2020_12_454_0_3"],
+        ),
+        # This param fails for the wrong reason -- because the 2019 data isn't available.
+        pytest.param(
+            ["True"],
+            [2020],
+            ["4270_2019_plant_total_11241"],
+            ["f1_steam_2020_12_454_0_3"],
+            marks=pytest.mark.xfail(
+                reason="EIA record year doesn't match FERC record year"
+            ),
+        ),
+        pytest.param(
+            ["True", "True"],
+            [2020, 2020],
+            ["4270_2020_plant_total_11241", "4270_2020_plant_total_11241"],
+            ["f1_steam_2020_12_454_0_3", "f1_steam_2020_12_159_0_3"],
+            marks=pytest.mark.xfail(reason="Duplicate EIA ids in training data"),
```
Made these xfail params, but they aren't guaranteed to fail for the right reasons. **Is there a way to ensure that they fail in a specific way?** The function this tests (`pudl.analysis.ferc1_eia_train.py::validate_override_fixes()`) throws a variety of AssertionErrors depending on whether certain conditionals are True or not. EX: 

```

    # Make sure there are no duplicate EIA id overrides
    logger.debug("Checking for duplicate override ids")
    assert (
        len(
            override_dups := only_overrides[
                only_overrides["record_id_eia_override_1"].duplicated(keep=False)
            ]
        )
        == 0
    ), f"Found record_id_eia_override_1 duplicates: \
    {override_dups.record_id_eia_override_1.unique()}"

    # Make sure the year in the EIA id overrides match the year in the report_year
    # column.
    logger.debug("Checking that year in override id matches report year")
    only_overrides = only_overrides.merge(
        ppl[["record_id_eia", "report_year"]].drop_duplicates(),
        left_on="record_id_eia_override_1",
        right_on="record_id_eia",
        how="left",
        suffixes=("", "_ppl"),
    )
    if (
        len(
            bad_eia_year := only_overrides["report_year"].compare(
                only_overrides["report_year_ppl"]
            )
        )
        > 0
    ):
        raise AssertionError(
            f"Found record_id_eia_override_1 values that don't correspond to the right \
            report year:\
            {[only_overrides.iloc[x].record_id_eia_override_1 for x in bad_eia_year.index]}"
        )
```

cmgosnell replied:
I believe you can add a specific [type of assertion into the xfail parameters](https://doc.pytest.org/en/latest/how-to/skipping.html#raises-parameter).


So in this case it would be something like 
```
marks=pytest.mark.xfail(reason="Duplicate EIA ids in training data",  raises=AssertionError)
```

(i think this would necessitate you converting the bare assert into a more specific assert, which is best practice anyway)

but i don't think you can tell it specifically which possible error in a specific module that it should trip on. I think that would require breaking out these defensive tests into smaller functions and running them individually. Which I believe is a good overall design because it makes things easier to test. But if i understand correctly, you want to be able to run specific training examples through the full validation process.

One thing to note: I *believe* some of these integration tests are run during the CI process as well at during the full nightly builds so this xfail may not always hold. I'll go look into that a little bit bc I'm not sure.

cmgosnell replied:
If you wanted to get super specific about why these tests are failing and when I think it would be much much easier to do if you broke out your `validate_override_fixes` into smaller chunks and tested each assert individually.

cmgosnell commented on test/integration/ferc1_eia_train_test.py:
```diff
@@ -0,0 +1,116 @@
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture(scope="module")
+def pudl_out(pudl_engine, pudl_datastore_fixture):
+    """A PUDL output object for use in CI."""
+    return PudlTabl(
+        pudl_engine,
+        ds=pudl_datastore_fixture,
+        freq="AS",
+        fill_tech_desc=False,
+        fill_net_gen=False,
+    )
```
I strongly recommend reusing another `PudlTabl` fixture.  I believe the `fast_out_annual` from `integration/output_tests.py` is probably the right one. It would probably be best to move `fast_out_annual` into `conftest` and convert it to `scope="session"`. I believe you could convert it to `scope="session"` and keep it in the output module.

cmgosnell commented on test/integration/ferc1_eia_train_test.py:
```diff
@@ -0,0 +1,116 @@
+    return pudl_out.utils_eia860()
+
+
+@pytest.fixture(scope="module")
+def plant_part_list(pudl_out):
+    """The plant_parts_eia output table."""
+    return pudl_out.plant_parts_eia().reset_index()
```
this is a nit we've been calling this the `plant_parts_eia` instead of `plant_part_list` in the code. its a hard habit to break colloquially but its basically `ppe` or `plant_parts_eia` in most places now.

cmgosnell commented on test/integration/ferc1_eia_train_test.py:
```diff
@@ -0,0 +1,116 @@
+            ["4270_2020_plant_total_11241"],
+            ["f1_steam_2020_12_454_0_3"],
+        ),
+        # This param fails for the wrong reason -- because the 2019 data isn't available.
+        pytest.param(
+            ["True"],
+            [2019],
+            ["4270_2020_plant_total_11241"],
+            ["f1_steam_2020_12_454_0_3"],
+            marks=pytest.mark.xfail(
+                reason="EIA record year doesn't match FERC record year"
+            ),
```
why are you adding the report_year here at all into this test? isn't the report year gleaned from either the `ppe` or the ferc plants table?

aesharpe replied:
this report_year column is how the validation function determines whether an EIA record was matched to a FERC record from a different year. Because the report_year column in the override spreadsheet comes from FERC.

cmgosnell commented on test/integration/ferc1_eia_train_test.py:
```diff
@@ -0,0 +1,116 @@
+        start_date=min(ferc1_eia.report_date),
+        end_date=max(ferc1_eia.report_date),
+    )
+    pytest.set_trace()
```
is this a relic of debugging?

aesharpe replied:
yeah!

cmgosnell commented on src/pudl/analysis/ferc1_eia_train.py:
```diff
@@ -0,0 +1,720 @@
+            how="left",
+            suffixes=("", "_ppl"),
+        )
+        # Now merge the utility_id_pudl from EIA in so that you can compare it with the
+        # utility_id_pudl from FERC that's already in the overrides
+        only_overrides = only_overrides.merge(
+            utils_eia860[["utility_id_eia", "utility_id_pudl"]].drop_duplicates(),
+            left_on="utility_id_eia_ppl",
+            right_on="utility_id_eia",
+            how="left",
+            suffixes=("", "_utils"),
+        )
```
if this is the only use of this `utils_eia860`, why not just get this info from the `ppe`? (trying to reduce the number of dfs flowing into your tests)

aesharpe replied:
That's a good idea

cmgosnell reviewed (commented):
I think you are definitely going in the right directions with these test! I left a few questions and suggestions. my main reaction was "wow there are a lot of inputs into `test_validate_override_fixes` and `validate_override_fixes`, which often means that something is trying to do too many things all at once. I think it would work as is, but debugging will be more challenging  with this setup versus breaking things out into smaller more testable chunks. I definitely won't die on this hill though if you can navigate it.

aesharpe commented:
Continued in #2502 to avoid all these weird file changes related to `dev` stuff.

e-belfer commented:
Closed in #2502
